providers/rpc/rpc.go

```diff
@@ -0,0 +1,369 @@
+	switch strings.ToLower(state) {
+	case "on", "off", "cycle":
+		rp := RequestPayload{
+			ID:     time.Now().UnixNano(),
+			Host:   p.Host,
+			Method: PowerSetMethod,
+			Params: PowerSetParams{
+				State: strings.ToLower(state),
+			},
+		}
+		resp, err := p.process(ctx, rp)
+		if err != nil {
+			return ok, err
+		}
+		if resp.Error != nil && resp.Error.Code != 0 {
+			return ok, fmt.Errorf("error from rpc consumer: %v", resp.Error)
+		}
+
+		return true, nil
+	}
+
+	return false, errors.New("requested power state is not supported")
+}
+
+// PowerStateGet gets the power state of a BMC machine.
+func (p *Provider) PowerStateGet(ctx context.Context) (state string, err error) {
+	rp := RequestPayload{
+		ID:     time.Now().UnixNano(),
+		Host:   p.Host,
+		Method: PowerGetMethod,
+	}
+	resp, err := p.process(ctx, rp)
+	if err != nil {
+		return "", err
+	}
+	if resp.Error != nil && resp.Error.Code != 0 {
+		return "", fmt.Errorf("error from rpc consumer: %v", resp.Error)
+	}
+
+	return resp.Result.(string), nil
+}
+
+// process is the main function for the roundtrip of rpc calls to the ConsumerURL.
+func (p *Provider) process(ctx context.Context, rp RequestPayload) (ResponsePayload, error) {
+	// 1. create the HTTP request.
+	// 2. create the signature payload.
+	// 3. sign the signature payload.
+	// 4. add signatures to the request as headers.
+	// 5. request/response round trip.
+	// 6. handle the response.
+	req, err := p.createRequest(ctx, rp)
+	if err != nil {
+		return ResponsePayload{}, err
+	}
+
+	// create the signature payload
+	reqBuf := new(bytes.Buffer)
+	if _, err := io.Copy(reqBuf, req.Body); err != nil {
+		return ResponsePayload{}, fmt.Errorf("failed to read request body: %w", err)
+	}
+	headersForSig := http.Header{}
+	for _, h := range p.Opts.Signature.IncludedPayloadHeaders {
+		if val := req.Header.Get(h); val != "" {
+			headersForSig.Add(h, val)
+		}
+	}
+	sigPay := createSignaturePayload(reqBuf.Bytes(), headersForSig)
+
+	// sign the signature payload
+	sigs, err := sign(sigPay, p.Opts.HMAC.Hashes, p.Opts.HMAC.PrefixSigDisabled)
+	if err != nil {
+		return ResponsePayload{}, err
+	}
+
+	// add signatures to the request as headers.
+	for algo, keys := range sigs {
+		if len(sigs) > 0 {
+			h := p.Opts.Signature.HeaderName
+			if !p.Opts.Signature.AppendAlgoToHeaderDisabled {
+				h = fmt.Sprintf("%s-%s", h, algo.ToShort())
+			}
+			req.Header.Add(h, strings.Join(keys, ","))
+		}
+	}
+
+	// request/response round trip.
+	kvs := requestKVS(req)
+	kvs = append(kvs, []interface{}{"host", p.Host, "method", rp.Method, "consumerURL", p.ConsumerURL}...)
+	if rp.Params != nil {
+		kvs = append(kvs, []interface{}{"params", rp.Params}...)
+	}
+
+	resp, err := p.Client.Do(req)
+	if err != nil {
+		p.Logger.Error(err, "failed to send rpc notification", kvs...)
+		return ResponsePayload{}, err
+	}
+	defer resp.Body.Close()
+	respBuf := new(bytes.Buffer)
+	if _, err := io.Copy(respBuf, resp.Body); err != nil {
```

Comment:
I think this may still be at risk of a DoS because `Copy` will read everything from `Body` into the buffer. I think, in theory, that could consume a CPU but also chew threw all available memory.

I think the only way to safely read is to chunk (https://pkg.go.dev/io#CopyN).

Reply:
ah yeah, fair. i'll use `CopyN` and move the copy below the content length check.